### PR TITLE
Add onRouteStart option

### PR DESCRIPTION
**Pull request by mjc1283**
@jordangarcia This adds the onRouteStart option.

### CODE REVIEW

File: src/Router.js [hunk 1]
```diff
@@ -160,6 +161,10 @@ export default class Router {
 
       this.__currentCanonicalPath = canonicalPath
 
+      if (this.onRouteStart && mode !== 'replace') {
+        this.onRouteStart();
```
Comment (jordangarcia) [inline src/Router.js]: I think it'd be good to supply some data to this function.  Like the `fromPath` and make this threads through properly with redirects, as well as the `startTime`.

Thoughts?

Comment (mjc1283) [inline src/Router.js]: Yea, good point, I'll pass `fromPath` and `startTime`. Could you clarify what you meant about redirects?

Comment (loganlinn) [inline src/Router.js]: Current canonical path might also be helpful for generalized onRouteStart functions. Would `ctx` be reasonable to be passed in as well?

Comment (jordangarcia) [inline src/Router.js]: Yeah if we have Route A redirecting to Route B:

- Route A start (should call onStart)
- Route A redirect
- Route A end (never called)
- Route B start (should not call onStart)
- Route B end (should call onEnd)

**Pull request merged**